File: DataStructure/LinkedList/24_swapPairs.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        
        first_node = head
        second_node = head.next

        # 妙啊 ！！！ 从后往前换！！
        first_node.next = self.swapPairs(second_node.next) 
        second_node.next = first_node

        return second_node


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        dummy = ListNode(-1)
        dummy.next = head

        prev_node = dummy

        while head and head.next:
            # 定义两个节点
            first_node = head
            second_node = head.next
            # 交换两个节点 换3条线
            prev_node.next = second_node
            first_node.next = second_node.next
            second_node.next = first_node

            prev_node = first_node
            head = first_node.next
        return dummy.next

File: DataStructure/LinkedList/24_swapPairs.cpp
#include <iostream>
#include <vector>
using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


// 这个思路应该是跟反转链表差不多
class Solution {
public:
    ListNode* reverseLinkList(ListNode* head) {
        ListNode* pre = NULL;
        ListNode* next;
        while (head) {
            next = head->next;
            head->next = pre;
            pre = head;
            head = next;
        }
        head = pre;
        return head;
    }
};

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == NULL || head->next == NULL)
            return head;
        ListNode* pre = NULL;
        ListNode* temp = head;
        ListNode* next = temp->next;
        head = head->next;
        // 这样就可以解决奇数个的问题了
        while (temp && next) {
            temp->next = next->next;
            next->next = temp;
            if (pre != NULL)
                pre->next = next;
            pre = temp;
            temp = temp->next;
            if (temp != NULL)
                next = temp->next;
        }
        return head;
    }
};
// 通过！！！

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == NULL || head->next == NULL)
            return head;
        ListNode *next = head->next;
        head->next = swapPairs(next->next);
        next->next = head;
        return next;
    }
};
// 一次通过